### PR TITLE
gpg: Log gpg output to LogWriter

### PR DESCRIPTION
**Pull request by doronbehar**
Continuation of https://github.com/gopasspw/gopass/pull/2576 .

### CODE REVIEW

**Comment by doronbehar** [comment]
Aish I ran `make test` and it should be better to run `make clean && make fulltest`... I'm getting a different error though:

```
>> BUILD, version = 1.15.13/b1cbd097, output = gopass [OK]
mkdir -p ~/.gnupg
>> TEST, "full-mode": race detector off
     ok         github.com/gopasspw/gopass      0.072s  coverage: 47.1% of statements
        github.com/gopasspw/gopass/helpers/changelog            coverage: 0.0% of statements
        github.com/gopasspw/gopass/helpers/man          coverage: 0.0% of statements
        github.com/gopasspw/gopass/helpers/msipkg               coverage: 0.0% of statements
        github.com/gopasspw/gopass/helpers/postrel              coverage: 0.0% of statements
        github.com/gopasspw/gopass/helpers/release              coverage: 0.0% of statements
     ok         github.com/gopasspw/gopass/internal/action      5.093s  coverage: 60.2% of statements
     ok         github.com/gopasspw/gopass/internal/action/exit 0.004s  coverage: 100.0% of statements
     ok         github.com/gopasspw/gopass/internal/action/pwgen        0.007s  coverage: 51.6% of statements
     ok         github.com/gopasspw/gopass/internal/audit       0.043s  coverage: 24.1% of statements
     ok         github.com/gopasspw/gopass/internal/backend     0.079s  coverage: 55.4% of statements
     ?          github.com/gopasspw/gopass/internal/backend/crypto      [no test files]
     ok         github.com/gopasspw/gopass/internal/backend/crypto/age  0.047s  coverage: 2.6% of statements
     ok         github.com/gopasspw/gopass/internal/backend/crypto/gpg  0.003s  coverage: 93.6% of statements
     ok         github.com/gopasspw/gopass/internal/backend/crypto/gpg/cli      0.113s  coverage: 60.8% of statements
     ok         github.com/gopasspw/gopass/internal/backend/crypto/gpg/colons   0.004s  coverage: 39.1% of statements
     ok         github.com/gopasspw/gopass/internal/backend/crypto/gpg/gpgconf  0.003s  coverage: 27.3% of statements
     ok         github.com/gopasspw/gopass/internal/backend/crypto/plain        0.004s  coverage: 85.3% of statements
     ?          github.com/gopasspw/gopass/internal/backend/storage     [no test files]
        github.com/gopasspw/gopass/internal/backend/storage/fossilfs            coverage: 0.0% of statements
     ok         github.com/gopasspw/gopass/internal/backend/storage/fs  0.325s  coverage: 63.6% of statements
     ok         github.com/gopasspw/gopass/internal/backend/storage/gitfs       0.088s  coverage: 48.1% of statements
     ok         github.com/gopasspw/gopass/internal/cache       1.108s  coverage: 78.7% of statements
        github.com/gopasspw/gopass/internal/cache/ghssh         coverage: 0.0% of statements
     ok         github.com/gopasspw/gopass/internal/completion/fish     0.004s  coverage: 100.0% of statements
     ok         github.com/gopasspw/gopass/internal/completion/zsh      0.003s  coverage: 100.0% of statements
     ok         github.com/gopasspw/gopass/internal/config      0.064s  coverage: 68.4% of statements
     ok         github.com/gopasspw/gopass/internal/config/legacy       0.044s  coverage: 64.8% of statements
     ok         github.com/gopasspw/gopass/internal/create      0.040s  coverage: 17.5% of statements
     ok         github.com/gopasspw/gopass/internal/cui 0.007s  coverage: 57.0% of statements
     ok         github.com/gopasspw/gopass/internal/diff        0.002s  coverage: 100.0% of statements
     ok         github.com/gopasspw/gopass/internal/editor      0.007s  coverage: 53.6% of statements
        github.com/gopasspw/gopass/internal/env         coverage: 0.0% of statements
        github.com/gopasspw/gopass/internal/hashsum             coverage: 0.0% of statements
        github.com/gopasspw/gopass/internal/hook                coverage: 0.0% of statements
     ok         github.com/gopasspw/gopass/internal/notify      0.003s  coverage: 56.1% of statements
     ok         github.com/gopasspw/gopass/internal/out 0.003s  coverage: 38.3% of statements
     ok         github.com/gopasspw/gopass/internal/pwschemes/argon2i   0.795s  coverage: 72.7% of statements
     ok         github.com/gopasspw/gopass/internal/pwschemes/argon2id  1.816s  coverage: 72.7% of statements
     ok         github.com/gopasspw/gopass/internal/pwschemes/bcrypt    0.698s  coverage: 75.0% of statements
        github.com/gopasspw/gopass/internal/queue               coverage: 0.0% of statements
     ok         github.com/gopasspw/gopass/internal/recipients  0.004s  coverage: 81.8% of statements
        github.com/gopasspw/gopass/internal/reminder            coverage: 0.0% of statements
     ok         github.com/gopasspw/gopass/internal/set 0.005s  coverage: 92.7% of statements
     ok         github.com/gopasspw/gopass/internal/store       0.003s  coverage: 100.0% of statements
] 2 / 2
❌ Failed to decode public key A3683834: public key "A3683834" not found
git initialized at /tmp/TestGit4192319314/001/sub
git configured at /tmp/TestGit4192319314/001/sub
--- FAIL: TestSet (0.00s)
    write_test.go:25:
                Error Trace:    /home/doron/repos/gopass/internal/store/leaf/write_test.go:25
                Error:          An error is expected but got nil.
                Test:           TestSet
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
[] ❌ Invalid recipients found: Invalid Recipients: john.doe: no valid keys (expired?),
[] ❌ Failed to decode public key john.doe: public key "john.doe" not found
[] ❌ Failed to decode public key john.doe: public key "john.doe" not found
[] ⚠ failed to check "foo/bar":
    [fatal] [non-fatal] Missing recipients on foo/bar: [john.doe]
Run fsck with the --decrypt flag to re-encrypt it automatically, or edit this secret yourself., Extra recipients on foo/bar: [0xDEADBEEF 0xFEEDBEEF]
Run fsck with the --decrypt flag to re-encrypt it automatically, or edit this secret yourself., Run fsck with the --decrypt flag to re-encrypt it automatically, or edit the secret foo/bar yourself.
failed to check "foo/baz":
    [fatal] [non-fatal] Missing recipients on foo/baz: [john.doe]
Run fsck with the --decrypt flag to re-encrypt it automatically, or edit this secret yourself., Extra recipients on foo/baz: [0xDEADBEEF 0xFEEDBEEF]
Run fsck with the --decrypt flag to re-encrypt it automatically, or edit this secret yourself., Run fsck with the --decrypt flag to re-encrypt it automatically, or edit the secret foo/baz yourself.
failed to check "foo/zab":
    [fatal] [non-fatal] Missing recipients on foo/zab: [john.doe]
Run fsck with the --decrypt flag to re-encrypt it automatically, or edit this secret yourself., Extra recipients on foo/zab: [0xDEADBEEF 0xFEEDBEEF]
Run fsck with the --decrypt flag to re-encrypt it automatically, or edit this secret yourself., Run fsck with the --decrypt flag to re-encrypt it automatically, or edit the secret foo/zab yourself.

⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
⚠ No owner key found. Make sure your key is fully trusted.
FAIL
coverage: 53.6% of statements
FAIL    github.com/gopasspw/gopass/internal/store/leaf  8.770s
FAIL
make: *** [Makefile:96: fulltest] Error 1
```

I'll look into this later today hopefully.

**Comment by doronbehar** [comment]
Fixed the tests. Hopefully now on CI as well.

**Comment by doronbehar** [comment]
Finally green :heavy_check_mark: :)

File: internal/backend/crypto/gpg/cli/encrypt.go [hunk 1]
```diff
@@ -25,25 +28,37 @@ func (g *GPG) Encrypt(ctx context.Context, plaintext []byte, recipients []string
...
+		return buf.Bytes(), errors.New("no valid and trusted recipients were found!")
+	}
 
 	cmd := exec.CommandContext(ctx, g.binary, args...)
 	cmd.Stdin = bytes.NewReader(plaintext)
-	// the encrypted blob is written to stdout
-	cmd.Stdout = buf
-	cmd.Stderr = os.Stderr
+	// the encrypted blob and errors are printed to the log file, and to stdout
+	cmd.Stdout = io.MultiWriter(buf, debug.LogWriter)
```
Comment (AnomalRoil) [inline internal/backend/crypto/gpg/cli/encrypt.go:60]: I wonder if we really want to write the encrypted blobs to the log file, this seems like it may contain sensitive data and information needed to debug gpg output should probably already be in `Stderr`, no?

File: internal/backend/crypto/gpg/cli/encrypt.go [hunk 2]
```diff
@@ -25,25 +28,37 @@ func (g *GPG) Encrypt(ctx context.Context, plaintext []byte, recipients []string
 		args = append(args, "--trust-model=always")
 	}
 
+	if len(recipients) == 0 {
+		debug.Log("recipients list is empty!")
```
Comment (AnomalRoil) [inline internal/backend/crypto/gpg/cli/encrypt.go:32]: shouldn't this warrant an early return?

File: internal/backend/crypto/gpg/cli/encrypt.go [hunk 3]
```diff
@@ -25,25 +28,37 @@ func (g *GPG) Encrypt(ctx context.Context, plaintext []byte, recipients []string
...
+			out.Printf(ctx, errmsg)
 
 			continue
 		}
+		debug.Log("adding recipient %s", r)
 		args = append(args, "--recipient", r)
 	}
-
 	buf := &bytes.Buffer{}
+	if len(badRecipients) == len(recipients) && len(recipients) != 0 {
```
Comment (AnomalRoil) [inline internal/backend/crypto/gpg/cli/encrypt.go:53]:
```suggestion
	if len(badRecipients) == len(recipients) {
```
`len(badRecipient)` is necessarily equal to 0 too if `len(recipients)` is 0, meaning this `if` would be true then, why wouldn't we want to take that return then?

I don't realize if there are valid usecases where `len(recipients) == 0` and we want to carry on with the `exec` below?

File: internal/backend/crypto/gpg/cli/encrypt.go [hunk 4]
```diff
@@ -25,25 +28,37 @@ func (g *GPG) Encrypt(ctx context.Context, plaintext []byte, recipients []string
...
+	}
+	var badRecipients []string
 	for _, r := range recipients {
 		kl, err := g.listKeys(ctx, "public", r)
 		if err != nil {
 			debug.Log("Failed to check key %s. Adding anyway. %s", err)
 		} else if len(kl.UseableKeys(gpg.IsAlwaysTrust(ctx))) < 1 {
-			out.Printf(ctx, "Not using invalid key %s for encryption. (Check its expiration date or its encryption capabilities.)", r)
+			badRecipients = append(badRecipients, r)
+			errmsg := fmt.Sprintf("Not using invalid key %s for encryption. Check its expiration date, its encryption capabilities and trust.", r)
```
Comment (AnomalRoil) [inline internal/backend/crypto/gpg/cli/encrypt.go:43]:
```suggestion
			errmsg := fmt.Sprintf("Not using invalid key %q for encryption. Check its expiration date, its encryption capabilities and trust.", r)
```

File: internal/backend/crypto/gpg/cli/encrypt.go [hunk 5]
```diff
@@ -25,25 +28,37 @@ func (g *GPG) Encrypt(ctx context.Context, plaintext []byte, recipients []string
...
+			debug.Log(errmsg)
+			out.Printf(ctx, errmsg)
```
Comment (AnomalRoil) [inline internal/backend/crypto/gpg/cli/encrypt.go:45]: Is the `debug.Log` really needed? I thought `out.Printf` was logging already:
https://github.com/gopasspw/gopass/blob/d00a8e1dbba5d3a8b4db19ee54f1734fa184dde4/internal/out/print.go#L47-L54

Is `ctxutil.IsHidden` returning `true` in this `Encrypt` method?

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 2 above)
Comment (doronbehar) [reply internal/backend/crypto/gpg/cli/encrypt.go:32]: > shouldn't this warrant an early return?

I'm not sure. See my comment to your other comment.

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 5 above)
Comment (doronbehar) [reply internal/backend/crypto/gpg/cli/encrypt.go:45]: > Is `ctxutil.IsHidden` returning `true` in this `Encrypt` method?

Good question... I think it depends on the context of the function calling this `Encrypt` method.. Could it be that when someone runs `gopass` from withing the web extension it is hidden? And that's why I experienced lack of useful logs when I tried to debug why it hadn't worked for me?

**Comment by doronbehar** [comment]
Don't understand the CI errors unfortuenatly :/ On other PRs it is red also...

**Comment by tcodes0** [comment]
@doronbehar hi there, you can find more information on the CI errors by looking for the name between parenthesis on this [link](https://golangci-lint.run/usage/linters); on this PR it's mostly the MND linter complaining about naked numbers in function calls. It's possible to ignore certain functions, like `os.OpenFile`, the docs should cover the ignore mechanism. Just passing by and saw your question :)

**Comment by doronbehar** [comment]
Thanks for taking the time to try to help @tcodes0 ! It's nice to understand now the errors, and I'm more convinced that these are not related to my PR, and that fixing them is out of scope for this PR. I pushed a rebase of my commits, but according to the [current latest commit](https://github.com/gopasspw/gopass/actions/runs/8958873719/job/24603626714?pr=2869) that shows the same errors in CI, I don't hope the errors will be fixed.

**Comment by doronbehar** [comment]
I'm still waiting for replies by @AnomalRoil for my old comments :pray:.

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 5 above)
Comment (dominikschulz) [reply internal/backend/crypto/gpg/cli/encrypt.go:45]: I think @AnomalRoil is right and we don't need the extra `debug.Log` statement, but also it doesn't hurt. So let's not block on this further.

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 1 above)
Comment (dominikschulz) [reply internal/backend/crypto/gpg/cli/encrypt.go:60]: I think writing encrypted blobs to the log is ok, but I wonder if it's useful?
Can you easily extract the encrypted blob from the log and operate on it with gpg?

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 2 above)
Comment (dominikschulz) [reply internal/backend/crypto/gpg/cli/encrypt.go:32]: This should be an early return. If that makes tests fail we need to fix them.

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 3 above)
Comment (dominikschulz) [reply internal/backend/crypto/gpg/cli/encrypt.go:53]: TestEncrypt is a simple (silly?) test that doesn't do much. If it fails because we introduce a new check feel free to update the expecation (i.e. change line 22 to `require.Error(t, err)`).

**Review by dominikschulz — COMMENTED** [review]
Almost there. Please address my comments and we can finally merge this. Sorry for the delays.

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 3 above)
Comment (doronbehar) [reply internal/backend/crypto/gpg/cli/encrypt.go:53]: > TestEncrypt is a simple (silly?) test that doesn't do much. If it fails because we introduce a new check feel free to update the expectation (i.e. change line 22 to `require.Error(t, err)`).

OK I did it. It is a bit even more silly now when I don't fully test the encrypt function. It'd be nice to fix that, but I'm not sure if that is possible when `g.binary` is set to `"true"`.

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 1 above)
Comment (doronbehar) [reply internal/backend/crypto/gpg/cli/encrypt.go:60]: > I think writing encrypted blobs to the log is ok, but I wonder if it's useful?

As I said in my previous comment: In the experience I had with `gopass-jsonapi listen`, I had available only the log opened in a file, and the stdout and stderr of the program were hidden in by the browser. Hence I would argue that it is useful to print as much information to the log.

> Can you easily extract the encrypted blob from the log and operate on it with gpg?

If we are talking about the user's gpg which has available the user's private keys etc, then yes. Other people shouldn't be able to decrypt these blobs if the user has their private keys kept private.

File: internal/backend/crypto/gpg/cli/encrypt.go (hunk 2 above)
Comment (doronbehar) [reply internal/backend/crypto/gpg/cli/encrypt.go:32]: > This should be an early return. If that makes tests fail we need to fix them.

Done.

**Comment by doronbehar** [comment]
> Almost there. Please address my comments and we can finally merge this. Sorry for the delays.

Thanks for coming back to this.

File: internal/backend/crypto/gpg/cli/gpg_windows_test.go [hunk 6]
```diff
@@ -17,7 +17,9 @@ func TestEncrypt(t *testing.T) {
 	g.binary = "rundll32"
 
 	_, err := g.Encrypt(ctx, []byte("foo"), nil)
-	require.NoError(t, err)
+
+	// No recipients are configured so it will fail
+	require.Error(t, err)
```
Comment (AnomalRoil) [inline internal/backend/crypto/gpg/cli/gpg_windows_test.go:22]: We might want to add a test with a recipient now to have a "positive" test, but that can be opened as an issue and done in another PR I think,

**Review by AnomalRoil — DISMISSED** [review]

**Comment by AnomalRoil** [comment]
The error in the Linux build seems like an error caused by the addition to the makefile, no?
```
--- FAIL: TestGPG (0.01s)
    gpg_test.go:30: 
        	Error Trace:	/home/runner/work/gopass/gopass/internal/backend/crypto/gpg/cli/gpg_test.go:30
        	Error:      	Received unexpected error:
        	            	exit status 2: tru::1:1720345348:0:3:1:5
        	            	|gpg: WARNING: unsafe permissions on homedir '/home/runner/.gnupg'
        	            	gpg: /home/runner/.gnupg/trustdb.gpg: trustdb created
        	            	gpg: [don't know]: invalid packet (ctb=00)
        	            	gpg: keydb_search_first failed: Invalid packet
        	Test:       	TestGPG
FAIL
``` 
not sure if it needs the right set of permissions for it to work or not.

**Comment by doronbehar** [comment]
> The error in the Linux build seems like an error caused by the addition to the makefile, no?

Could be.. `make fulltest` worked for me locally. I removed that last commit now. Let's see how CI goes now.

File: internal/backend/crypto/gpg/cli/gpg_windows_test.go (hunk 6 above)
Comment (doronbehar) [reply internal/backend/crypto/gpg/cli/gpg_windows_test.go:22]: > We might want to add a test with a recipient now to have a "positive" test, but that can be opened as an issue and done in another PR I think,

I wanted to do that, but I didn't find a method to add a recipient to that context / gpg in that test...

**Comment by doronbehar** [comment]
OK now only the old GOLANGCI-LINT errors are present.

**Review by dominikschulz — APPROVED** [review]

**Comment by doronbehar** [comment]
Thanks!
